core/src/test/java/org/apache/iceberg/rest/requests/TestCreateNamespaceRequest.java

```diff
@@ -0,0 +1,108 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.rest.requests;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.Map;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.util.JsonUtil;
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+
+public class TestCreateNamespaceRequest {
+
+  private static final ObjectMapper MAPPER = JsonUtil.withRESTComponents();
+
+  private static final String[] ALL_FIELDS_FROM_SPEC = { "namespace", "properties" };
+
+  /* Values used to fill in request fields */
+  private static final Namespace NAMESPACE = Namespace.of("accounting", "tax");
+  private static final Map<String, String> PROPERTIES = ImmutableMap.of("owner", "Hank");
+
+  /* JSON of requests */
+  private static final String JSON_REQUEST =
+      "{\"namespace\":[\"accounting\",\"tax\"],\"properties\":{\"owner\":\"Hank\"}}";
+  private static final String JSON_REQUEST_EMPTY_PROPERTIES =
+      "{\"namespace\":[\"accounting\",\"tax\"],\"properties\":{}}";
+
+  @Test
+  public void testDeserialize() throws JsonProcessingException {
+    CreateNamespaceRequest expected = CreateNamespaceRequest.builder()
+        .withNamespace(NAMESPACE)
+        .withProperties(PROPERTIES)
+        .build();
+
+    CreateNamespaceRequest actual = MAPPER.readValue(JSON_REQUEST, CreateNamespaceRequest.class);
+
+    assertEquals(expected, actual);
+    assertHasOnlyKnownFields(expected);
+    assertHasOnlyKnownFields(actual);
+  }
+
+  @Test
+  public void testDeserializeEmptyProperties() throws JsonProcessingException {
+    CreateNamespaceRequest expected = CreateNamespaceRequest.builder()
+        .withNamespace(NAMESPACE)
+        .build();
+
+    CreateNamespaceRequest actual = MAPPER.readValue(JSON_REQUEST_EMPTY_PROPERTIES, CreateNamespaceRequest.class);
+
+    assertEquals(expected, actual);
+    assertHasOnlyKnownFields(expected);
+    assertHasOnlyKnownFields(actual);
+  }
+
+  @Test
+  public void testSerialize() throws JsonProcessingException {
+    CreateNamespaceRequest req = CreateNamespaceRequest.builder()
+        .withNamespace(NAMESPACE)
+        .withProperties(PROPERTIES)
+        .build();
+
+    Assertions.assertThat(MAPPER.writeValueAsString(req))
+        .isEqualTo(JSON_REQUEST);
+  }
+
+  @Test
+  public void testSerializeEmptyProperties() throws JsonProcessingException {
+    CreateNamespaceRequest req = CreateNamespaceRequest.builder()
+        .withNamespace(NAMESPACE)
+        .build();
+
+    Assertions.assertThat(MAPPER.writeValueAsString(req))
+        .isEqualTo(JSON_REQUEST_EMPTY_PROPERTIES);
+  }
+
+  private static void assertEquals(CreateNamespaceRequest expected, CreateNamespaceRequest actual) {
+    Assertions.assertThat(actual.namespace()).isEqualTo(expected.namespace());
```

Comment:
just wondering whether it would make sense to have `CreateNamespaceRequest` just implement `equals()` rather than having to do it in test code?

Reply:
Yeah I agree. I changed it to simply implement `equals` (and `hashCode`).

Reply:
So implementing `equals` lead to having to jump through a lot of hoops to ensure the builder couldn't create values with `null` where not accessed (and lazily instantiate a number of things).

Long story short, it lead to writing a lot of code that was more to get tests to pass than for the class to work. So I've extracted out an abstract base test class, and then it required the implementation of `assertEquals`. It also still checks that it knows of all fields. So if somebody adds a field, it will be caught by the test.

I can show you further, but it was definitely uglier due to comparing `null` with empty collections (`null` coming from deserialization and empty collections coming from the builder by default).

Reply:
Another comment about it: https://github.com/apache/iceberg/pull/4037#discussion_r808449734